### PR TITLE
Map dataset on off in phase maker

### PR DESCRIPTION
**Pull request by MRegeard**
**Description**

This pull request modify the PhaseBackgroundMaker to support a MapDatasetOnOff dataset. This is useful to be able to produce pulsed map. 
I also modify the test_phase file to be able to test the run method of the maker in the case of a MapDatasetOnOff.
Because of the current handling of the phase in the data (namely adding column(s) to the DL3 event file), I also added a column_name ("PHASE" by default) in the init to be able to choose the phase column if there are multiples. For this I put it in the init of the maker, but I am not sure that it should be here or in the run method directly. 
I was also wandering whether it is a good idea to put a method that check if the phase column is non-zeros. This thought is mainly motivated by the fact that in HESS DL3 event file, a PHASE column exist by default but is filled with zeros. Is this a relevant feature or do we consider that it is the responsibility of the user to check it. 
I haven't checked what happens if you give a column name that does not exist in the event file. My guess would be that an error occur, but that's probably something that I can check. 

**Dear reviewer**

### CODE REVIEW

File: gammapy/makers/background/phase.py [hunk 1]
```diff
@@ -19,32 +24,39 @@ class PhaseBackgroundMaker(Maker):
...
     def __str__(self):
         s = self.__class__.__name__
         s += f"\n{self.on_phase}"
         s += f"\n{self.off_phase}"
+        s += f"\n{self.column_name}"
         return s
 
     @staticmethod
-    def _make_counts(dataset, observation, phases):
+    def _make_counts(self, dataset, observation, phases):
```
Comment (AtreyeeS) [inline gammapy/makers/background/phase.py:46]: pass the column name instead of accessing `self`

File: gammapy/makers/background/phase.py [hunk 2]
```diff
@@ -100,20 +112,37 @@ def run(self, dataset, observation):
...
 
-        acceptance = RegionNDMap.from_geom(geom=dataset.counts.geom)
-        acceptance.data = np.sum([_[1] - _[0] for _ in self.on_phase])
+        if isinstance(dataset, SpectrumDataset):
+            acceptance = RegionNDMap.from_geom(geom=dataset.counts.geom)
+            acceptance.data = np.sum([_[1] - _[0] for _ in self.on_phase])
 
-        acceptance_off = RegionNDMap.from_geom(geom=dataset.counts.geom)
-        acceptance_off.data = np.sum([_[1] - _[0] for _ in self.off_phase])
+            acceptance_off = RegionNDMap.from_geom(geom=dataset.counts.geom)
```
Comment (AtreyeeS) [inline gammapy/makers/background/phase.py:119]: Same as before. No need for the `if`, create a `Map` directly

File: gammapy/makers/background/phase.py [hunk 3]
```diff
@@ -100,20 +112,37 @@ def run(self, dataset, observation):
...
 
-        dataset_on_off = SpectrumDatasetOnOff.from_spectrum_dataset(
-            dataset=dataset,
-            counts_off=counts_off,
-            acceptance=acceptance,
-            acceptance_off=acceptance_off,
-        )
-        dataset_on_off.counts = counts
-        return dataset_on_off
+            dataset_on_off = SpectrumDatasetOnOff.from_spectrum_dataset(
```
Comment (AtreyeeS) [inline gammapy/makers/background/phase.py:122]: Maybe simplify a bit by
```
dataset_on_off = MapDatasetOnOff.from_spectrum_dataset(
             dataset=dataset,
             counts_off=counts_off,
             acceptance=acceptance,
             acceptance_off=acceptance_off,
         )

if `dataset._geom.is_region:
     dataset_on_off = dataset_on_off.to_spectrum_dataset(dataset._geom.region)

return dataset_on_off
```

File: gammapy/makers/background/phase.py [hunk 4]
```diff
@@ -1,8 +1,13 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import numpy as np
 from gammapy.data import EventList
-from gammapy.datasets import SpectrumDatasetOnOff
-from gammapy.maps import RegionNDMap
+from gammapy.datasets import (
```
Comment (AtreyeeS) [inline gammapy/makers/background/phase.py:4]: these imports are not necessary

**Review by AtreyeeS — CHANGES_REQUESTED** [review]
Thank you @MRegeard !

This looks good. I have left some inline comments to decrease the lines of code a bit.
I like the idea of passing the name of the phase column - the present logic is good, it should happen on `init`

I do not think that we should worry about zero or other invalid values in the phase column at present. Creating this column is rather specific to analysis of a specific pulsar, and at this point it is the users' responsibility to ensure that the phasing is done correctly.

File: gammapy/makers/background/phase.py (hunk 3 above)
Comment (MRegeard) [reply gammapy/makers/background/phase.py:119]: Hi @AtreyeeS,
I have made all the modifications that you suggest apart from this one. 
If I understand correctly what you suggest, it means that in both case (SpectrumonOff or MapOnOff), the user has to pass a MapDataset to the maker? (I guessed that it MapDatasetOnOff.from_map_dataset([...]) and not MapDatasetOnOff.from_spectrum_dataset([...])).
If I understand correctly, I'm not sure that it is very intuitive for the user to build a MapDataset and then to have the maker transform it into a SpectrumDataset if the user wants to do a SpectrumDataset from the start. 
Let me know

Comment (AtreyeeS) [reply gammapy/makers/background/phase.py:119]: Hi @MRegeard ! Yes indeed, it should be `MapDatasetOnOff.from_map_dataset` as you correctly pointed out. Sorry! 

The user does not need to create a `MapDataset` explicitly.
`MapDatasetOfOff.xyz()` functionalities are available on `SpectrumDataset` due to inheritance. So, `MapDatasetOnOff.from_map_dataset` will create a MapDatasetOnOff. Now, if dataset was initially a spectral one to start off with, gammapy will change to a spectral datasetonoff

Comment (AtreyeeS) [reply gammapy/makers/background/phase.py:119]: Thinking a bit about this, I think it is better to change 

```
if dataset._geom.is_region:
     dataset_on_off = dataset_on_off.to_spectrum_dataset(dataset._geom.region)
```

to 

```
if isInstance(dataset, SpectrumDataset):
    dataset_on_off = dataset_on_off.to_spectrum_dataset(dataset._geom.region)
```

Comment (MRegeard) [reply gammapy/makers/background/phase.py]: Oh I see, I always forget about inheritance...
Changes were made !

**Review by AtreyeeS — DISMISSED** [review]
Thanks @MRegeard ! No further comments from my side

File: gammapy/makers/background/phase.py [hunk 5]
```diff
@@ -19,32 +19,36 @@ class PhaseBackgroundMaker(Maker):
         on-phase defined by the two edges of each interval (edges are excluded)
     off_phase : `tuple` or list of tuples
         off-phase defined by the two edges of each interval (edges are excluded)
+    column_name : `str`
+        The name of the column in the event file from which the phase informations are extracted
```
Comment (registerrier) [inline gammapy/makers/background/phase.py:23]:
```suggestion
        The name of the column in the event file from which the phase informations are extracted. Default is 'PHASE'.
```

File: gammapy/makers/background/phase.py [hunk 6]
```diff
@@ -19,32 +19,36 @@ class PhaseBackgroundMaker(Maker):
...
+    def __init__(self, on_phase, off_phase, column_name="PHASE"):
         self.on_phase = self._check_intervals(on_phase)
         self.off_phase = self._check_intervals(off_phase)
+        self.column_name = column_name
 
     def __str__(self):
         s = self.__class__.__name__
         s += f"\n{self.on_phase}"
         s += f"\n{self.off_phase}"
+        s += f"\n{self.column_name}"
```
Comment (registerrier) [inline gammapy/makers/background/phase.py:37]: Note that the `__str__` method could be more talkative.

File: gammapy/makers/background/phase.py [hunk 7]
```diff
@@ -19,32 +19,36 @@ class PhaseBackgroundMaker(Maker):
...
 
-    def __init__(self, on_phase, off_phase):
+    def __init__(self, on_phase, off_phase, column_name="PHASE"):
         self.on_phase = self._check_intervals(on_phase)
         self.off_phase = self._check_intervals(off_phase)
+        self.column_name = column_name
 
     def __str__(self):
         s = self.__class__.__name__
         s += f"\n{self.on_phase}"
```
Comment (registerrier) [inline gammapy/makers/background/phase.py:35]:
```suggestion
        s += f"\nON phase interval : {self.on_phase}"
```

File: gammapy/makers/background/phase.py (hunk 6 above)
Comment (registerrier) [inline gammapy/makers/background/phase.py:37]:
```suggestion
        s += f"\nOn phase interval : {self.on_phase}"
        s += f"\nOff phase interval : {self.off_phase}"
        s += f"\nPhase column name : {self.column_name}"
```

**Review by registerrier — APPROVED** [review]
Thanks @MRegeard . This looks good.

I have left some inline suggestions. 

Once this is merged, the associated tutorial should be updated.

File: gammapy/makers/background/phase.py [hunk 8]
```diff
@@ -19,32 +19,36 @@ class PhaseBackgroundMaker(Maker):
         on-phase defined by the two edges of each interval (edges are excluded)
     off_phase : `tuple` or list of tuples
         off-phase defined by the two edges of each interval (edges are excluded)
+    column_name : `str`
```
Comment (adonath) [inline gammapy/makers/background/phase.py:22]: Rename to the even more explicit `phase_column_name` or `column_name_phase`?

**Comment by MRegeard** [comment]
@registerrier changes done !
Yes the update of the tutorial is planned.

File: gammapy/makers/background/phase.py (hunk 8 above)
Comment (MRegeard) [reply gammapy/makers/background/phase.py:22]: Yes I guess it would be better.

**Review by registerrier — APPROVED** [review]
Thanks @MRegeard !

**Comment by registerrier** [comment]
The CI fail is unrelated. Merging now.

**Pull request merged**